WordPress/gutenberg
# Writing flow: A11Y enhancements for useTabNav

## What?


This is a fairly complex refactor of writing flow for accessibility. Specifically the tabbing functionality in useTabNav.

## Why?


See the following discussion for details. Essentially, Shift+Tab/Tab keys should be reserved for navigating a blocks content, not triggering actions such as toolbar focus or sidebar jump (to be addressed in another PR).

https://github.com/WordPress/gutenberg/discussions/38311

## How?


1. Modified isFormField() function to finally be compatible with contenteditable.
2. Refactored useTabNav() in writing flow to do the following.
  1. Fixed one of the if conditionals to ensure elements exist in the same block before allowing focus to move. E.g., You should be able to press Tab to get from Upload to Media Library buttons for the image block, but not pressing Tab to get from paragraph to heading blocks.
  2. The Tab key with the Shift+Tab key should navigate block content reliably.

## Testing Instructions





1. Open a post or page.
2. Insert an image block.
3. Tab to the different buttons and even in to the sidebar.
4. Now use Shift+Tab back to the block and through the available focus candidates. Once you run out of previous focus candidates, you should land safely in the block toolbar.
5. This can also be tested with the table block as it has multiple text input fields after the table is created.

## Screenshots or screencast
